### ISSUE TITLE
Deriving from Protocol results in error when not all required members are implemented

### ISSUE THREAD

**Issue by rphes** [microsoft/pyright#3730]
**Describe the bug**
According to the [mypy docs](https://mypy.readthedocs.io/en/latest/more_types.html#mixin-classes), a preferred way of doing mixins is to annotate the `self` argument of mixin methods with a `Protocol` that specifies which members should be available to the mixin. The issue with this is that methods annotated this way can no longer access the members of the mixin they belong to. [This answer](https://stackoverflow.com/questions/51930339/how-do-i-correctly-add-type-hints-to-mixin-classes/70907644#70907644) on stack overflow suggests to just inherit from the `Protocol` instead.

In mypy, this works, but in pyright it results in an error saying not all required members are implemented. [PEP 544 states](https://peps.python.org/pep-0544/#merging-and-extending-protocols) that classes inheriting from a `Protocol` should get "downgraded" to a regular ABC, which has me thinking pyright should only error when the mixin is instantiated, not when it is defined.

**To Reproduce**
Adapted the stack overflow answer:

```python
from typing import Protocol


class HasValueProtocol(Protocol):
    @property
    def value(self) -> int: ...


class MultiplicationMixin(HasValueProtocol): # error in pyright, no error in mypy
    def multiply(self, m: int) -> int:
        return self.value * m


class AdditionMixin(HasValueProtocol): # same as above
    def add(self, b: int) -> int:
        return self.value + b


class MyClass(MultiplicationMixin, AdditionMixin):
    @property
    def value(self) -> int:
       return 5

mixin = AdditionMixin() # not an error in both pyright and mypy
```

The errors:

```
test.py
  test.py:8:7 - error: Class derives from one or more protocol classes but does not implement all required members
    Member "value" is declared in protocol class "HasValueProtocol" (reportGeneralTypeIssues)
  test.py:13:7 - error: Class derives from one or more protocol classes but does not implement all required members
    Member "value" is declared in protocol class "HasValueProtocol" (reportGeneralTypeIssues)
2 errors, 0 warnings, 0 information
```

**Expected behavior**

- I should be able to inherit from the `Protocol` class without error
- I should not be able to instantiate an incomplete mixin class

Does this make sense? I might be missing something.

**Comment by erictraut** [comment]
Pyright is working as intended here. If you provide no implementation for the `value` property in the protocol, then a class that derives from the protocol must supply an implementation. Pyright interprets a function body of `...` as "no implementation". If you change the implementation of `HasValueProtocol.value` to `return 0`, the error goes away.

**Comment by erictraut** [comment]
Alternatively, if you want `MultiplicationMixin` and `AdditionMixin` to act like abstract classes, you should explicitly declare them as protocols themselves.

```python
class MultiplicationMixin(HasValueProtocol, Protocol): ...
```

**Comment by rphes** [comment]
Appreciate the quick response!

Would you then disagree that the passage from PEP 544 I mentioned before:

> Subclassing a protocol class would not turn the subclass into a protocol unless it also has typing.Protocol as an explicit base class. Without this base, the class is “downgraded” to a regular ABC that cannot be used with structural subtyping.

implies the two mixin classes should be treated as ABCs?

I still can't quite grasp what the right solution for annotating mixin classes depending on some common member would be.

**Comment by DylanLukes** [comment]
+1 to @rphes here. The current behavior does not seems to be in line with PEP 544. 

At the very least, _if the subclass is explicitly marked as an ABC_ this should not cause an error.
